### PR TITLE
Issue1983 boiler plant staging setpoints

### PR DESCRIPTION
**Pull request by karthikeyad-pnnl**
Pull request to pull all files for boiler plant staging setpoint subsequences.

### CODE REVIEW

**Comment by karthikeyad-pnnl** [comment]
@milicag : This pull request is ready for review. It contains all the files for the staging setpoint subsequences.

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+    "Design boiler capacities vector"
+    annotation(Dialog(tab="General", group="Boiler plant configuration parameters"));
+
+  parameter Real boiFirMin[nBoi](
+    final unit="1",
+    final displayUnit="1")
+    "Boiler minimum firing ratio"
+    annotation(Dialog(tab="General", group="Boiler plant configuration parameters"));
+
+  parameter Real bMinPriPumSpeSta[nSta](
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo:44]: what is b? Maybe a typo and it should have been `boi`?

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+    final unit="1",
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo:45]: please use fill function as described here `https://obc.lbl.gov/specification/cdl.html#sec-dec-par`. For example:

```
  CDL.Interfaces.RealInput uChiWatIsoVal[nChi](
    final min=fill(0, nChi),
    final max=fill(1, nChi),
    final unit=fill("1", nChi))
    "Chilled water isolation valve position"
    annotation (Placement(transformation(extent={{-840,-220},{-800,-180}}),
      iconTransformation(extent={{-140,-50},{-100,-10}})));
```

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput VHotWat_flow(
+    final quantity="VolumeFlowRate",
+    final unit="m3/s",
+    final displayUnit="m3/s")
+    "Measured hot water flow rate"
+    annotation (Placement(transformation(extent={{-440,190},{-400,230}}),
+      iconTransformation(extent={{-140,90},{-100,130}})));
+
+  Buildings.Controls.OBC.CDL.Interfaces.RealInput VMinSet_flow[nSta](
+    final unit="m3/s",
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo:283]: Just as a reminder please apply the `fill` function consistently for any vector variable.

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+    final delDesCapConBoi=delDesCapConBoi,
+    final sigDif=sigDif,
+    final delBypVal=delBypVal,
+    final bypValClo=bypValClo,
+    final TCirDif=TCirDif,
+    final delTRetDif=delTRetDif,
+    final dTemp=dTemp,
+    final TDif=TDif,
+    final delFaiCon=delFaiCon,
+    bMinPriPumSpeSta=bMinPriPumSpeSta)
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo:401]: looks like you forgot `final` here, also see if `b` needs to be `boi` - related to one of the previous comments.

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+    "Staging matrix with stage as row index and boiler as column index"
+    annotation(Dialog(tab="General", group="Boiler plant configuration parameters"));
+
+  parameter Integer iniSta = 1
+    "Initial boiler plant stage"
+    annotation(Dialog(tab="General", group="Boiler plant configuration parameters"));
+
+  parameter Real boiDesCap[nBoi](
+    final unit="W",
+    final displayUnit="W",
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/SetpointController.mo:33]: In the latest edits to the documentation that are now in a PR Michael would like us to not use `final` with `displayUnit`. I am planning to soon do a fix for the chiller plant package.

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/Subsequences/EfficiencyCondition.mo [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+block EfficiencyCondition
+  "Efficiency condition used in staging up and down"
+
+  parameter Integer nSta = 5
+    "Number of stages in the boiler plant";
+
+  parameter Real fraNonConBoi(
+    final min=0,
+    final unit="1",
+    final displayUnit="1") = 0.9
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/Subsequences/EfficiencyCondition.mo:11]: Just a reminder to remove `final`. Now that the code is formatted nicely, a simple bulk rename in a text editor can be used.

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/Subsequences/FailsafeCondition.mo [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    final quantity="ThermodynamicTemperature")
+    "Measured hot water supply temperature"
+    annotation (Placement(transformation(extent={{-160,-20},{-120,20}}),
+      iconTransformation(extent={{-140,-20},{-100,20}})));
+
+  Buildings.Controls.OBC.CDL.Interfaces.BooleanOutput yFaiCon
+    "Failsafe condition for boiler staging"
+    annotation (Placement(transformation(extent={{120,-20},{160,20}}),
+      iconTransformation(extent={{100,-20},{140,20}})));
+
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/Subsequences/FailsafeCondition.mo:50]: 1 empty line is enough probably, although I personally don't have preferences on this.

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/Subsequences/Validation/BoilerIndices.mo [hunk 8]
```diff
@@ -0,0 +1,143 @@
+within Buildings.Controls.OBC.ASHRAE.PrimarySystem.BoilerPlant.Staging.SetPoints.Subsequences.Validation;
+model BoilerIndices
+  "Validates extraction of boiler indices in a given stage"
+
+  Buildings.Controls.OBC.ASHRAE.PrimarySystem.BoilerPlant.Staging.SetPoints.Subsequences.BoilerIndices
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/Subsequences/Validation/BoilerIndices.mo:5]: Michael decided to not mandate the location of the tested instances, but still recommend they be grouped together.

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/Subsequences/Validation/EfficiencyCondition.mo [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+          -160},{111,-12}}, color={255,0,255}));
+  connect(booPul.y, edg.u)
+    annotation (Line(points={{172,-160},{178,-160}}, color={255,0,255}));
+  connect(edg.y, effCon2.uStaChaProEnd) annotation (Line(points={{202,-160},{251,
+          -160},{251,-12}}, color={255,0,255}));
+  connect(con11.y, effCon2.uCapReq) annotation (Line(points={{172,120},{240,120},
+          {240,9},{248,9}}, color={0,0,127}));
+  connect(con12.y, effCon2.VHotWat_flow)
+    annotation (Line(points={{172,0},{248,0}}, color={0,0,127}));
+  annotation(Icon(coordinateSystem(preserveAspectRatio=false,
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/Subsequences/Validation/EfficiencyCondition.mo:240]: Please don't forget to add revisions section.

Comment (milicag) [reply Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/Subsequences/Validation/EfficiencyCondition.mo:240]: This applies to all blocks and validation models, please make sure they all have it.

File: Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/package.order [hunk 10]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (milicag) [inline Buildings/Controls/OBC/ASHRAE/PrimarySystem/BoilerPlant/Staging/SetPoints/package.order:1]: remove empty line

**Comment by milicag** [comment]
@karthikeyad-pnnl Please note change in PR title for clarity.

**Comment by karthikeyad-pnnl** [comment]
@milicag : This pull request is ready for review.

**Comment by karthikeyad-pnnl** [comment]
Closing pull request since all sequences will be pulled through #2193.

**Pull request closed without merging**